src/compartment.c

```diff
@@ -0,0 +1,526 @@
+                            break;
+                        }
+                        default:
+                            assert(false);
+                    }
+                    found += 1;
+                    printf("Found `main` func at %p.\n", (void*) new_comp->entry_point);
+                }
+                else
+                {
+                    for (size_t i = 0; i < MAX_INTERCEPT_COUNT; ++i)
+                    {
+                        if (!strcmp(comp_intercept_funcs[i].func_name, &comp_strtb[curr_sym.st_name]))
+                        {
+                            comp_add_intercept(new_comp, curr_sym.st_value, comp_intercept_funcs[i]);
+                            found += 1;
+                            break;
+                        }
+                    }
+                }
+                if (found == to_find)
+                {
+                    break;
+                }
+            }
+            free(comp_symtb);
+            free(comp_strtb);
+        }
+        // TODO still need relas check or consider only static executables?
+        else if (comp_symtb_hdr.sh_type == SHT_RELA) // TODO change name && consider SH_REL
+        {
+            if (comp_symtb_hdr.sh_info == 0) // TODO better identify the plt relocation section
+            {
+                continue;
+            }
+
+            new_comp->relas_cnt = comp_symtb_hdr.sh_size / sizeof(Elf64_Rela);
+            new_comp->relas = (uintptr_t*) malloc(new_comp->relas_cnt * sizeof(uintptr_t));
+            Elf64_Rela* comp_relas = (Elf64_Rela*) malloc(comp_symtb_hdr.sh_size);
+            pread_res = pread((int) new_comp->fd, comp_relas, comp_symtb_hdr.sh_size, comp_symtb_hdr.sh_offset);
+            assert(pread_res != -1);
+
+            for (size_t j = 0; j < new_comp->relas_cnt; ++j)
+            {
+                new_comp->relas[j] = comp_relas[j].r_offset;
+                /*comp_relas[j].r_offset += new_comp->base;*/
+                /*uintptr_t old_plt_val = (uintptr_t) *((void**) comp_relas[j].r_offset);*/
+                /*old_plt_val += new_comp->base;*/
+            }
+
+            free(comp_relas);
+        }
+        else
+        {
+            continue;
+        }
+    }
+    /*assert(found == to_find);*/
+
+    comp_register_ddc(new_comp);
+    comp_print(new_comp);
+    return new_comp;
+}
+
+void
+comp_register_ddc(struct Compartment* new_comp)
+{
+    void* __capability new_ddc = cheri_address_set(manager_ddc, new_comp->base);
+    new_ddc = cheri_bounds_set(new_ddc, new_comp->size + new_comp->scratch_mem_size + new_comp->scratch_mem_stack_size);
+    // TODO bounds double-check
+    new_comp->ddc = new_ddc;
+}
+
+void
+comp_add_intercept(struct Compartment* new_comp, uintptr_t intercept_target, struct func_intercept intercept_data)
+{
+    printf("Found `%s` func at %p.\n", intercept_data.func_name, (void*) intercept_target);
+    int offset = (intercept_data.redirect_func - intercept_target) / 4;
```

Comment:
> As I directly control those 24 bits, shouldn't int be sufficient to hold that data?

In portable C `int` is only required to have 16 bits, but most systems these days (including Morello) give it 32.

However, the problem here is about the truncation _to_ `int` in C, not the instruction encoding. This is a general C concern, not directly related to the 24/26-bit integer. For example, if `intercept_data.redirect_func = 0x00005678000000a8` and `intercept-target = 0x0000123400000000`, then:

```
int offset = (0x00005678000000a8 - 0x0000123400000000) / 4;
assert(offset < (1 << 27));
```

i.e.


```
int offset = 0x11110000002a;   // UB! Typically truncates to 0x0000002a (42).
assert(offset < (1 << 27));   // UB continues, but this typically succeeds with offset=42.
...  // Typically, `b <PC+42>` will be emitted.
```

If you instead compute that `offset` with 64 bits (and unsigned), then the same assertion works fine. It'll still fail the assertion if it's out of range, but it won't quietly do the wrong thing, and won't have any undefined behaviour. (At least, not _that_ undefined behaviour.)

Reply:
Alright, I understand this example. So when we are computing things related to addresses and offsets to pass to the instructions, we should be using an unsigned 64-bit type, and you suggest the best is `ptraddr_t`. Did I understand that correctly?

I have gone ahead and update my working copy (which should closely follow this PR) with this change for the instruction synthesis.

Reply:
> Did I understand that correctly?

Yes, though it isn't about addresses and offsets specifically, but most arithmetic involving most types!

In general: take care not to truncate in intermediate values. (In C, this can even be implicit.) Sometimes you can't avoid it — most operators can overflow and for many types that would lead to undefined behaviour — but you can choose what assumptions you want to make depending on the situation.

In this case, I left a review comment because you'd `assert`ed and so clearly _intended_ to catch the overflow case. (I might have left one anyway since using a bigger type fixes this for minimal cost, but that would have a different weighting.)